executor/pipeline/fetch.rs

```diff
@@ -95,125 +121,144 @@ fn execute_fetch_some(
+            Err(FetchExecutionError::Unimplemented { description: "List attributes are not available yet." })
+        }
+    }
+}
+
+fn execute_single_attribute(
+    snapshot: Arc<impl ReadableSnapshot + 'static>,
+    thing_manager: Arc<ThingManager>,
+    row: MaybeOwnedRow<'_>,
+    position: &VariablePosition,
+    attribute_type: &AttributeType<'static>,
+) -> Result<DocumentNode, FetchExecutionError> {
+    let variable_value = row.get(*position).as_reference();
+    match variable_value {
+        VariableValue::Empty => Ok(DocumentNode::Leaf(DocumentLeaf::Empty)),
+        VariableValue::Thing(Thing::Entity(entity)) => {
+            execute_attribute_single(entity, attribute_type.clone(), snapshot, thing_manager)
+                .map(|leaf| DocumentNode::Leaf(leaf))
         }
+        VariableValue::Thing(Thing::Relation(relation)) => {
+            execute_attribute_single(relation, attribute_type.clone(), snapshot, thing_manager)
+                .map(|leaf| DocumentNode::Leaf(leaf))
+        }
+        VariableValue::Thing(Thing::Attribute(_)) => Err(FetchExecutionError::FetchAttributesOfAttribute {}),
+        VariableValue::Type(_) => Err(FetchExecutionError::FetchAttributesOfType {}),
+        VariableValue::Value(_) => Err(FetchExecutionError::FetchAttributesOfValue {}),
+        VariableValue::ThingList(_) | VariableValue::ValueList(_) => Err(FetchExecutionError::FetchAttributesOfList {}),
     }
 }
 
+fn execute_single_function(
+    snapshot: Arc<impl ReadableSnapshot + 'static>,
+    thing_manager: Arc<ThingManager>,
+    parameters: Arc<ParameterRegistry>,
+    functions_registry: Arc<ExecutableFunctionRegistry>,
+    row: MaybeOwnedRow<'_>,
+    mut interrupt: ExecutionInterrupt,
+    variable_positions: &HashMap<Variable, VariablePosition>,
+    function: &ExecutableFunction,
+) -> Result<DocumentNode, FetchExecutionError> {
+    let (mut pattern_executor, execution_context) = prepare_single_function_execution(
+        snapshot,
+        thing_manager,
+        parameters,
+        functions_registry.clone(),
+        variable_positions,
+        row,
+        function,
+    )?;
+    let mut tabled_functions = TabledFunctions::new(functions_registry.clone());
+    let mut suspend_points = Vec::new();
+
+    let batch = exactly_one_or_return_err!(
+        pattern_executor
+            .compute_next_batch(&execution_context, &mut interrupt, &mut tabled_functions, &mut suspend_points)
+            .map_err(|err| FetchExecutionError::ReadExecution { typedb_source: Box::new(err) })?,
+        FetchExecutionError::FetchSingleFunctionNotSingle { func_name: "func".to_string() } // TODO: Can we get function name here?
+    );
+
+    let batch = match batch {
+        Some(batch) => batch,
+        None => return Ok(DocumentNode::Leaf(DocumentLeaf::Empty)),
+    };
+
+    let mut row_iter = batch.into_iter();
```

Comment:
I suspect it might be neat/nice to have another type of `BatchRowIterator` that you build that returns you a Row iterator instead of  Batch iterator -> Row iterator within a batch? So you don't have to deal with this manual stuff...

Reply:
I left a todo here to move forward this week as I didn't quickly understand how to make it in a good way with the lack of understanding of the whole local codebase.

Reply:
ok